### PR TITLE
Features/453635 create independent tabs

### PR DESCRIPTION
**Pull request by molochiy**

### CODE REVIEW

File: src/app/applicationbuilder.cpp [hunk 1]
```diff
@@ -213,12 +213,12 @@ void ApplicationBuilder::createUiActions(MainWindow *pMainWindow)
 {
     UserAction *lActionOpen = new UserAction(tr("&Open Repository..."), mMainWindow);
     // TASK: add implementation for opening existing repository
-    mActionManager->add(FileMenuGroup, lActionOpen);
+    mActionManager->add(FileMenuGroup, lActionOpen);
```
Comment (Houd1ny) [inline src/app/applicationbuilder.cpp]: remove spaces

File: src/app/ui/contentfortabworkplace.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    ui->setupUi(this);    
+    QSizePolicy sizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+    setSizePolicy(sizePolicy);
+    setMouseTracking(false);
+    setAutoFillBackground(false);
+    setDockNestingEnabled(true);
+    setDockOptions(QMainWindow::AllowNestedDocks|QMainWindow::AllowTabbedDocks|QMainWindow::AnimatedDocks);
+
+    uiHistoryTree = new HistoryTree(this);
+    uiHistoryTree->setMinimumSize(QSize(250, 38));
```
Comment (Houd1ny) [inline src/app/ui/contentfortabworkplace.cpp]: add coments for magic numbers
38 looks funny

File: src/app/ui/contentfortabworkplace.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    dockWidgetContents_5 = new RevisionView();
+    uiHistoryTree->setWidget(dockWidgetContents_5);
+    addDockWidget(static_cast<Qt::DockWidgetArea>(8), uiHistoryTree);
+
+
+    uiEditorView = new EditorView(this);
+    uiEditorView->setMinimumSize(QSize(461, 38));
+    dockWidgetContents = new QWidget();
+    uiEditorView->setWidget(dockWidgetContents);
+    addDockWidget(static_cast<Qt::DockWidgetArea>(8), uiEditorView);
```
Comment (Houd1ny) [inline src/app/ui/contentfortabworkplace.cpp:33]: add comment 
I don`t get wat is going on here
I mean static_castQt::DockWidgetArea(8)

File: src/app/ui/contentfortabworkplace.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+
+    uiFileView = new DockFileView(this);
+    sizePolicy.setHeightForWidth(uiFileView->sizePolicy().hasHeightForWidth());
+    uiFileView->setSizePolicy(sizePolicy);
+    uiFileView->setMinimumSize(QSize(250, 38));
+    uiFileView->setBaseSize(QSize(500, 370));
+    uiFileView->setFloating(false);    
+    dockWidgetContents_4 = new FileView();
```
Comment (Houd1ny) [inline src/app/ui/contentfortabworkplace.cpp]: codding standart
_4 ...
and wat does it mean forth
should there be list or smth like that

File: src/app/ui/contentfortabworkplace.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    uiFileView->setWidget(dockWidgetContents_4);
+    addDockWidget(static_cast<Qt::DockWidgetArea>(1), uiFileView);
+
+    uiRevisionTable = new RevisionTable(this);
+    uiRevisionTable->setMinimumSize(QSize(461, 38));
+    dockWidgetContents_8 = new QWidget();
```
Comment (Houd1ny) [inline src/app/ui/contentfortabworkplace.cpp]: dockWidgetContents_8
wat 8?
are u kidding me?

Comment (Houd1ny) [inline src/app/ui/contentfortabworkplace.cpp]: where is parent?

File: src/app/ui/contentfortabworkplace.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
+#ifndef CONTENTFORTABWORKPLACE_H
+#define CONTENTFORTABWORKPLACE_H
+
```
Comment (Houd1ny) [inline src/app/ui/contentfortabworkplace.h:3]: do u really need all this includes?

File: src/app/ui/contentfortabworkplace.h [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+class ContentForTabWorkplace : public QMainWindow
+{
+    Q_OBJECT
+
+public:
+    explicit ContentForTabWorkplace(QWidget *parent = 0);
+    ~ContentForTabWorkplace();
+
+    HistoryTree *uiHistoryTree;
```
Comment (Houd1ny) [inline src/app/ui/contentfortabworkplace.h]: = NULL?

Comment (Houd1ny) [inline src/app/ui/contentfortabworkplace.h]: public members
should be private!

File: src/app/ui/customtabbar.cpp [hunk 8]
```diff
@@ -20,11 +22,11 @@ CustomTabBar::~CustomTabBar()
     delete ui;
 }
 
-void CustomTabBar::slotAddTab(int index, QString name)
+void CustomTabBar::slotAddNewWorkplace(int index, QString name)
 {
     Q_UNUSED(index);
     //TASK: check whether everything is  ok with index - it is not set
-    QTabBar::addTab(name);
+    QTabWidget::addTab(new ContentForTabWorkplace, name);
```
Comment (Houd1ny) [inline src/app/ui/customtabbar.cpp]: should be there parent?

File: src/app/ui/mainwindow.cpp [hunk 9]
```diff
@@ -32,27 +33,19 @@ MainWindow::MainWindow(QWidget *parent) :
     TrayIcon->setContextMenu(trayMenu);
 
     // TASK: creation of PageManager should be performed by ApplicationBuilder
-    mPageManager = new PageManager(this);
+    mTabsAPI = new TabsAPI();
```
Comment (Houd1ny) [inline src/app/ui/mainwindow.cpp]: parent?

File: src/app/ui/mainwindow.cpp [hunk 10]
```diff
@@ -32,27 +33,19 @@ MainWindow::MainWindow(QWidget *parent) :
...
-//    connect(mPageManager, SIGNAL(currentPageChanged(int)), ui->uiFileView, SLOT(slotSetPage(int)));
-    connect(mPageManager, SIGNAL(currentPageChanged(int)), ui->uiRevisionTable, SLOT(slotSetPage(int)));
-    connect(mPageManager, SIGNAL(currentPageChanged(int)), ui->uiConsole, SLOT(slotSetPage(int)));
-    connect(mPageManager, SIGNAL(currentPageChanged(int)), ui->uiHistoryTree, SLOT(slotSetPage(int)));
-    connect(mPageManager, SIGNAL(currentPageChanged(int)), ui->uiEditorView, SLOT(slotSetPage(int)));
-    connect(mTabBar, SIGNAL(tabMoved(int,int)), mPageManager, SLOT(slotTabMoved(int,int)));
+    this->setCentralWidget(mTabBar);
+
+    connect(mTabsAPI, SIGNAL(newWorkplaceAdded(int, QString)), mTabBar, SLOT(slotAddNewWorkplace(int,QString)), Qt::UniqueConnection);
+    connect(mTabBar, SIGNAL(tabCloseRequested(int)), mTabsAPI, SLOT(slotRemovePage(int)));
```
Comment (Houd1ny) [inline src/app/ui/mainwindow.cpp]: prev line unique connection
he not specified
wat type of connection u need?

File: src/app/ui/mainwindow.cpp [hunk 11]
```diff
@@ -92,12 +85,14 @@ void MainWindow::slotQuit()
...
 }
 
-void MainWindow::slotAddPage()
+void MainWindow::slotAddNewWorkplace()
 {
-    static int a = 0;
-    QString lTabName = "Name" + QString::number(a++);
-    // TASK: page should be added in TabBuilder class
-    mPageManager->slotAddPage(lTabName);
+    static int amountOpenedTabs = 0;
```
Comment (Houd1ny) [inline src/app/ui/mainwindow.cpp]: static?????
should be member
if no prove why should be static

File: src/app/ui/tabsapi.cpp [hunk 12]
```diff
@@ -0,0 +1,116 @@
+#include "tabsapi.h"
+#include <QDebug>
```
Comment (Houd1ny) [inline src/app/ui/tabsapi.cpp]: if u are no using qDebug
remove

File: src/app/ui/tabsapi.cpp [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+{
+    emit currentWorkplaceChanged(getTabID(index));
+}
+
+
+int TabsAPI::getTabID()
+{
+    if(mTabHash.isEmpty())
+    {
+    mTabHash.insert(mTabHash.count(), mTabHash.count());
```
Comment (Houd1ny) [inline src/app/ui/tabsapi.cpp]: \t
codding standart

File: src/app/ui/tabsapi.cpp [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    else
+    {
+
+        QList<int> lKeys = mTabHash.keys();
```
Comment (Houd1ny) [inline src/app/ui/tabsapi.cpp]: wat is the aim of algorithm?
add comments.
it looks strange

File: src/app/ui/tabsapi.cpp [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    return mTabHash.value(mTabHash.count());
+
+}
+
+int TabsAPI::getTabID(int index)
+{
+    //TASK: fixme
+    return mTabHash.key(index, -1);
+//    for(QMap<int,int>::iterator iter = mPageHash.begin(); iter!=mPageHash.end(); iter++)
```
Comment (Houd1ny) [inline src/app/ui/tabsapi.cpp]: remove commented code

File: src/app/ui/tabsapi.cpp [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+//        if(iter.value() == index)
+//           return iter.key();
+}
+
+int TabsAPI::getActiveIndex()
+{
+
```
Comment (Houd1ny) [inline src/app/ui/tabsapi.cpp]: too many new lines
why two additional "\n"

File: src/app/ui/tabsapi.cpp [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+{
+    emit newWorkplaceAdded(getTabID(), lName);
+}
+
+void TabsAPI::slotRemoveWorkplace(int index)
+{
+    QMap<int,int>::iterator iter=mTabHash.begin();
+    for(; iter!=mTabHash.end(); iter++)
+    {
+        if(iter.value() == index)
```
Comment (Houd1ny) [inline src/app/ui/tabsapi.cpp]: codding standart

File: src/app/ui/tabsapi.cpp [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+            break;
+    }
+
+    for(QMap<int,int>::iterator iter = mTabHash.begin(); iter!=mTabHash.end(); iter++)
+    {
+        if(iter.value() > index)
+            iter.value()--;
```
Comment (Houd1ny) [inline src/app/ui/tabsapi.cpp]: magic omg
- codding standart

File: src/app/ui/tabsapi.cpp [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    qDebug()<<"iter key "<<iter.key();
+    mTabHash.remove(iter.key());
+}
+
+void TabsAPI::slotTabMoved(int from, int to)
+{
+    int lFrom = 0, lTo = 0;
```
Comment (Houd1ny) [inline src/app/ui/tabsapi.cpp]: codding standart

File: src/app/ui/tabsapi.h [hunk 20]
```diff
@@ -0,0 +1,57 @@
+#ifndef PAGEMANAGER_H
+#define PAGEMANAGER_H
```
Comment (Houd1ny) [inline src/app/ui/tabsapi.h]: no tab api define
could be funny errors man)

File: src/app/ui/tabsapi.h [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+/*!
+ * \brief The PageManager class handles all pages for repositories.
+ */
+class TabsAPI : public QObject
+{
+    Q_OBJECT
+public:
+    explicit TabsAPI(QObject *parent = 0);
+    int getTabID();
+    int getTabID(int);
```
Comment (Houd1ny) [inline src/app/ui/tabsapi.h]: comments?

File: src/app/ui/viewsettingpage.cpp [hunk 22]
```diff
@@ -77,7 +77,7 @@ void ViewSettingPage::slotBtnOpen()
 //    }
 }
 
-void ViewSettingPage::slotFileViewChanged(QVariant pValue)
+/*void ViewSettingPage::slotFileViewChanged(QVariant pValue)
```
Comment (Houd1ny) [inline src/app/ui/viewsettingpage.cpp:80]: commented code?
remove

File: src/app/ui/viewsettingpage.h [hunk 23]
```diff
@@ -51,12 +51,12 @@ private slots:
     void slotBtnOpen();
 
 public slots:
-    void slotFileViewChanged(QVariant pValue);
+   /* void slotFileViewChanged(QVariant pValue);
```
Comment (Houd1ny) [inline src/app/ui/viewsettingpage.h:54]: remove unused code

File: src/app/ui/contentfortabworkplace.h (hunk 6 above)
Comment (molochiy) [inline src/app/ui/contentfortabworkplace.h:3]: Yes, I do

File: src/app/ui/tabsapi.cpp (hunk 14 above)
Comment (molochiy) [inline src/app/ui/tabsapi.cpp]: its not my code
idk that doing this algorithm

File: src/app/ui/viewsettingpage.cpp (hunk 22 above)
Comment (molochiy) [inline src/app/ui/viewsettingpage.cpp:80]: its not working settings, mb it will be need

File: src/app/ui/contentfortabworkplace.h (hunk 6 above)
Comment (Houd1ny) [inline src/app/ui/contentfortabworkplace.h:3]: http://stackoverflow.com/questions/4757565/c-forward-declaration
HistoryTree *uiHistoryTree;
-    RevisionView *uiHistoryTreeContents;
-    EditorView *uiEditorView;
-    FileView *uiFileViewContents;
-    DockFileView *uiFileView;
-    RevisionTable *uiRevisionTable;
-    DockConsole *uiConsole;
-    ConsoleView *uiConsoleContents;
  that`s just pointers

File: src/app/ui/tabsapi.cpp (hunk 14 above)
Comment (Houd1ny) [inline src/app/ui/tabsapi.cpp]: It`s yout code
Becouse it`s green
you add it as new code

Comment (molochiy) [inline src/app/ui/tabsapi.cpp]: i rename file and class from PageManager to TabsAPI, through this code is green

File: src/app/ui/contentfortabworkplace.h (hunk 6 above)
Comment (molochiy) [inline src/app/ui/contentfortabworkplace.h:3]: thx for this

**Pull request merged**